cloud/services/natgateways/natgateways.go

```diff
@@ -0,0 +1,93 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package natgateways
+
+import (
+	"context"
+
+	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2019-06-01/network"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/pkg/errors"
+
+	azure "sigs.k8s.io/cluster-api-provider-azure/cloud"
+)
+
+// Reconcile gets/creates/updates a nat gateway.
+func (s *Service) Reconcile(ctx context.Context) error {
+	if !s.Scope.Vnet().IsManaged(s.Scope.ClusterName()) {
+		s.Scope.V(4).Info("Skipping nat gateways reconcile in custom vnet mode")
+		return nil
+	}
+
+	for _, natGatewaySpec := range s.Scope.NatGatewaySpecs() {
+		existingNatGateway, err := s.Get(ctx, s.Scope.ResourceGroup(), natGatewaySpec.Name)
+		if !azure.ResourceNotFound(err) {
+			if err != nil {
+				return errors.Wrapf(err, "failed to get nat gateway %s in %s", natGatewaySpec.Name, s.Scope.ResourceGroup())
+			}
+
+			// nat gateway already exists
+			// currently don't support:
```

Comment:
If you intend for this iteration so support only one nat gateway, please change the logic here and around `NatGatewaySpecs()` to return a single element, not a slice. Right now this overwrites the name/id below will be overwritten on each iteration, but because we hardcode one nat gateway spec, it happens to work.